Mathlib/GroupTheory/GroupAction/Primitive.lean

```diff
@@ -0,0 +1,279 @@
+that has no fixed point acts pretransitively -/
+class _root_.AddAction.IsQuasiPreprimitive
+    [AddGroup G] [AddAction G X] extends AddAction.IsPretransitive G X : Prop where
+  isPretransitive_of_normal :
+    ∀ {N : AddSubgroup G} [N.Normal], AddAction.fixedPoints N X ≠ .univ →
+      AddAction.IsPretransitive N X
+
+/-- An  action of a group is quasipreprimitive if any normal subgroup
+that has no fixed point acts pretransitively -/
+@[to_additive]
+class IsQuasiPreprimitive [Group G] [MulAction G X] extends IsPretransitive G X : Prop where
+  isPretransitive_of_normal :
+    ∀ {N : Subgroup G} [N.Normal], fixedPoints N X ≠ .univ → IsPretransitive N X
+
+variable {G X}
+
+@[to_additive]
+theorem IsPreprimitive.subsingleton_or_eq_univ_of_isBlock
+    [SMul G X] [IsPreprimitive G X] {B : Set X} (hB : IsBlock G B) :
+    B.Subsingleton ∨ B = .univ :=
+  isTrivialBlock_of_isBlock hB
+
+@[to_additive (attr := nontriviality)]
+theorem isPreprimitive_of_subsingleton [SMul G X] [Nonempty G] [Subsingleton X] :
+    IsPreprimitive G X where
+  exists_smul_eq (x y) := by
+    use Classical.arbitrary G
+    rw [eq_iff_true_of_subsingleton]
+    trivial
+  isTrivialBlock_of_isBlock B := by
+    left
+    exact Set.subsingleton_of_subsingleton
+
+variable [Group G] [MulAction G X]
+
+open scoped BigOperators Pointwise
+
+/-- If the action is pretransitive, then the trivial blocks condition implies preprimitivity
+(based condition) -/
+@[to_additive
+"If the action is pretransitive, then the trivial blocks condition implies preprimitivity
+(based condition)"]
+theorem isPreprimitive_of_isTrivialBlock_base [IsPretransitive G X] (a : X)
+    (H : ∀ {B : Set X} (_ : a ∈ B) (_ : IsBlock G B), IsTrivialBlock B) :
+    IsPreprimitive G X where
+  isTrivialBlock_of_isBlock {B} hB := by
+    obtain rfl | ⟨b, hb⟩ := B.eq_empty_or_nonempty
+    · simp [IsTrivialBlock]
+    · obtain ⟨g, hg⟩ := exists_smul_eq G b a
+      rw [← IsTrivialBlock.smul_iff g]
+      apply H _ (hB.translate g)
+      rw [← hg]
+      use b
+
+/-- If the action is not trivial, then the trivial blocks condition implies preprimitivity
+(pretransitivity is automatic) (based condition) -/
+@[to_additive
+  "If the action is not trivial, then the trivial blocks condition implies preprimitivity
+  (pretransitivity is automatic) (based condition)"]
+theorem isPreprimitive_of_isTrivialBlock_of_not_mem_fixedPoints {a : X} (ha : a ∉ fixedPoints G X)
+    (H : ∀ ⦃B : Set X⦄, a ∈ B → IsBlock G B → IsTrivialBlock B) :
+    IsPreprimitive G X :=
+  have : IsPretransitive G X := by
+    rw [isPretransitive_iff_base a]
+    cases' H (mem_orbit_self a) (IsBlock.orbit a) with H H
+    · exfalso; apply ha
+      rw [Set.subsingleton_iff_singleton (mem_orbit_self a)] at H
+      simp only [mem_fixedPoints]
+      intro g
+      rw [← Set.mem_singleton_iff]; rw [← H]
+      exact mem_orbit a g
+    · intro x; rw [← MulAction.mem_orbit_iff, H]; exact Set.mem_univ x
```

Comment:
Shouldn't you make this a lemma about `IsPretransitive` in an earlier file?

Reply:
finally, I'd rather not, because it's really about the consequence of having trivial blocks only, which is primitivity, so it wouldn't really make sense in another file.